Updating ThesesFR translator

eonm-abes opened the pull request:
Hi all I updated the ThesesFR.js translator.

__Change log :__

* The translator uses a RDF file to get Metadata. This RDF file is more precise and up to date than embedded metadata in the html page
* The translators is able to import multiples items from organization and person pages
* Supervisors are considered as contributors of the final work
* The mention "Thèses de doctorat" (defended thesis) or "Thèse en préparation" (thesis in preparation) has been added to the type field. This change was requested by theses.fr users.
* More info has been added in the extra field (jury members, laboratory)
* The mention "Les données de Theses.fr sont sous licence Etalab" has been added in the authorization field. (__Can this field be used for licensing purposes ?__)

It's the first time I contribute to update a translator so feedback are welcome. I had some trouble to lint the code with eslint I hope there is no more issues.

A few questions nonetheless :

* I use `const` as often as possible to avoid reassignment. Is this a good practice in the context of translators ? Should I use `let` instead ?
* I use arrow functions. Can it causes compatibility issues ?

Regards

TesteFixe commented:
Nice Work!

adam3smith commented:
@dstillman -- do you have an opinion on `const`? See e.g.
https://github.com/zotero/translators/pull/2150/files#diff-ae76dd6444254ff5be2191eeb6d4189cR51
for a respresentative example. 
I don't see a downside, but it's not something we've been doing (let alone enforcing) in translators otherwise.

adam3smith commented on ThesesFR.js:
```diff
@@ -79,24 +85,80 @@ function doWeb(doc, url) {
+		const title = ZU.xpathText(xmlDoc, '//dc:title', ns);
+		
+		// The case should't exists : preve;nts crashes while importing multiple records
+		if (!title) return;
```
Let's take this out please or replace it with a `throw` -- if an item unexpectedly doesn't import, we would _want_ an error to be thrown rather than to continue silently, letting the user assume all selected items have been imported.

adam3smith commented on ThesesFR.js:
```diff
@@ -79,24 +85,80 @@ function doWeb(doc, url) {
+		});
+
+		const abstractNotes = ZU.xpath(xmlDoc, '//dcterms:abstract', ns);
+		newItem.abstractNote = abstractNotes.length > 0 ? abstractNotes[0].textContent : undefined;
```
Why no just use `ZU.xpathText()` here? (and similarly in other cases below)

eonm-abes replied:
My xpath skills was a bit rusty. I failed to get the first element of the abstract (and other nodes) with pure xpath.

Originally I tried to use this expression `//dcterms:abstract[1]` with `xpathText()` function, but this expression was wrong in my context. It should have been `(//dcterms:abstract)[1]`.

So I used a workaround by mixing xpath and javascript which is not the best implementation. Now that I have found the right xpath expression I'll fix the code to use `xpathText()` function.

dstillman replied:
(Note that you can also use `text(docOrElem, selector[, index])` or `attr(docOrElem, selector, attribute[, index])` with CSS selectors instead of using XPaths.)

eonm-abes replied:
I already fixed the code by using xpath (2c91144) and `Z.xpathText()` but that's a good trick to know.

adam3smith commented on ThesesFR.js:
```diff
@@ -79,24 +85,80 @@ function doWeb(doc, url) {
+		newItem.rights = 'Les données de Theses.fr sont sous licence Etalab';
+
+		// Keep extra information such as laboratory, graduate schools, etc.
+		newItem.extra = Array.from(doc.getElementsByClassName('donnees-ombre')).map((description) => {
```
This is better suited for a note than `extra`

eonm-abes replied:
In this case I'll use a note instead

adam3smith commented on ThesesFR.js:
no more PDFs? That seems like it'd be useful

eonm-abes replied:
@adam3smith I agree with you, It would be very useful. I tried to implement it my self but I encountered multiple issues with PDF attachments :
- Most PDF files are around 2Go or more (Theses.fr doesn't enforce PDF size). It might cause real performance issues when you have to import multiple PDF files at once. In my case my web browser slows down or freezes. Over time PDF files get bigger and bigger on theses.fr
- Finding full text PDF on theses.fr is a hard task. Currently there is a special path in the URL `/document` that redirects user to the full text PDF if it exists. But this URL redirects either to :
   - The PDF file on theses.fr (https://www.theses.fr/2011PA040275/document)
   - The PDF file on another website (University website, Open archive, ...) (https://www.theses.fr/2008ECDL0023/document). In this case we are fetching resources from 3rd party websites without user consent. Furthermore **there is no way to know if the URL is pointing to a third-party website before being redirected**. Currently half of open access PDF are stored on external websites. 
  - A web page containing the PDF file. In this case we have to parse each pages individually to figure out where the PDF is stored. It implies to make multiple HTTP calls to find and download the PDF file. (https://www.theses.fr/2020REN20023/document).
  - A restricted access web page (https://www.theses.fr/2019LIL1R054/document)

It seems like the previous implementation is also broken.

To implement it I need more time and maybe help. I'm also concerned about user consent.

adam3smith replied:
It's certainly fine to leave this out, I was just checking.

FWIW, I'm not super concerned about user consent to retrieve PDFs from 3rd party academic websites -- Zotero's unpaywall integration does this, too. The fact that we'd only be able to successfully import PDFs from 1 and 2 of your four scenarios with reasonable effort is also fine: that's why we check the mimeType.

However, we don't have a way to check for file size limits in translators and I certainly don't want to auto-import 2GB document, so if this is a regular occurrence, that'd indeed be a problem.

adam3smith commented on ThesesFR.js:
```diff
@@ -36,41 +36,47 @@
-		return "multiple";
+	// Match against a results page or a Ph. D/supervisor/organization page which might contains multiple records e.g.
+	// http://www.theses.fr/fr/?q=zotero
+	// http://www.theses.fr/fr/154750417
```
cool -- please add a test for this

adam3smith commented:
I don't quite understand this question:

> The mention "Les données de Theses.fr sont sous licence Etalab" has been added in the authorization field. (Can this field be used for licensing purposes ?) 

are you asking whether the `rights` field is the correct place for licensing information? If so, the answer is yes.

dstillman commented:
> @dstillman -- do you have an opinion on const? […] I don't see a downside, but it's not something we've been doing (let alone enforcing) in translators otherwise.

We generally don't use `const` anywhere in Zotero beyond actual constants (e.g., all caps variables). It's more verbose without adding much value beyond `let` (particularly in small scopes, where it's just noise). There's no need to reject translators that use `const`, but for brevity and consistency we prefer it not be used.

(We also tend to use `var` at the top level of functions to differentiate variables shared across multiple scopes from block-level `let`, as well as to allow for quicker `try` debugging, but we certainly don't enforce that.)

eonm-abes commented:
Thank you @adam3smith  and @dstillman for your comments and feedback. I updated the code accordingly to your advice.

@adam3smith my question was indeed about the right place to put licensing information. I updated the code to keep only the license name (Etalab).

@dstillman As suggested I replaced `const` with `let`.

It should be good now.

adam3smith commented on ThesesFR.js:
```diff
@@ -230,16 +247,19 @@ var testCases = [
 			   		}
 				],
-				"extra": "Sous la direction de  Maarten Boonekamp.  Soutenue le 19-09-2016,à l'Université Paris-Saclay (ComUE) , dans le cadre de   École doctorale Particules, Hadrons, Énergie et Noyau : Instrumentation, Imagerie, Cosmos et Simulation (Orsay, Essonne ; 2015-....) , en partenariat avec  Département de physique des particules (Gif-sur-Yvette, Essonne)   (laboratoire)  ,  Centre européen pour l ...
 				"rights": "Les données de Theses.fr sont sous licence Etalab",
```
doesn't look like you updated the tests after changing rights

adam3smith commented on ThesesFR.js:
```diff
@@ -108,53 +107,61 @@ function scrape(doc, url) {
-		newItem.extra = Array.from(doc.getElementsByClassName('donnees-ombre')).map((description) => {
-			return Array.from(description.getElementsByTagName('p')).map(description => description.textContent.replace(/\n/g, ' '));
+		if (notePrepa) {
+			newItem.notes.push({ description: notePrepa });
```
This should be `{note: notePrepa}` as per https://www.zotero.org/support/dev/translators/coding#notes 
Same below

adam3smith commented on ThesesFR.js:
```diff
@@ -108,53 +107,61 @@ function scrape(doc, url) {
 		}).join(' ');
 
+		if (note) {
+			newItem.notes.push({ description: note });
```
ditto

adam3smith commented:
Hi -- sorry, I know I said we'd be done, but I'm just realizing that the combination of target and detect function is significantly too loose: The translator detects any page as a thesis, including search start pages (http://theses.fr/fr/personnes/) and documentation (http://theses.fr/fr/personnes/). This isn't your bug, it's already the case in the current version, but it's... not good.

I don't think you can easily address this in the target, so will probably require something on the page itself -- one of the metatags might make most sense, e.g. DC.type?

Would you mind tightening this up a bit more?

eonm-abes commented:
No problem. This is my fault, I forget to check the target regex when I updated the translator.

I come up with this regex `^https?:\/\/(www\.)?theses\.fr\/([a-z]{2}\/)?((s\d+|\d{4}.{8}|\d{8}X|\d{9})(?!\.(rdf|xml)$)|(sujets\/\?q=|\?q=))(?!.*&format=(json|xml))`. What do you think of it ?

This regex should exclude search start pages, .xml and .rdf pages and API pages.

More explanation and example here : https://regex101.com/r/IIr1ak/1

adam3smith commented:
Thanks so much. This looks great!
